_posts/2018-12-06-pmreorder-basics.md

```diff
@@ -0,0 +1,318 @@
+static void
+list_insert(struct list_root *root, node_id node, int value)
+{
+	struct list_node *new = NODE_PTR(root, node);
+
+	new->next = root->head;
+	pmem_persist(&new->next, sizeof(node));
+
+	root->head = node;
+	pmem_persist(&root->head, sizeof(root->head));
+
+	new->value = value;
+	pmem_persist(&new->value, sizeof(value));
+}
+{% endhighlight %}
+
+Full implementation and explanation you can find in [pmdk repository][pmreorderExampleLink].
+
+At first, the next value in the node is set and persisted.
+Then node is linked at the beginning of the list, its value is set
+and both persisted separately.
+
+When you run this code under pmemcheck you will get the following result:
+ 
+{% highlight sh %}
+==22161== Number of stores nod made persistent: 0
+==22161== ERROR SUMMARY: 0 errors
+{% endhighlight %}
+
+It means that all stores are persisted properly, as we expected.
+
+Now it is time to check the list under pmreorder tool.
+To do that you need to consider when your list is consistent
+and on that basics write consistency checker function.
+
+Follow the next chapters to get to know more details about that.
+
+
+### Checker
+
+Consistency checker is just a binary/function that defines
+conditions necessary to fulfill your consistency assumptions.
+It should return 0 if the state is consistent or 1 if it isn't.
+
+In case of our list with three elements:
+
+{% highlight C linenos %}
+ // params: root, node_id, value
+list_insert(r, 5, 55);
+list_insert(r, 3, 33);
+list_insert(r, 6, 66);
+{% endhighlight %}
+
+we would like to be sure that if node (5, 3, 6) is linked to the list
+then its value is set properly (55, 33, 66).
+If the node will be linked to the list but its value field will have
+an improper number then you should recognize this situation as inconsistent
+and return 1.
+
+For example:
+{% highlight C linenos %}
+static int
+check_consistency(struct list_root *root)
+{
+	struct list_node *node1 = NODE_PTR(root, root->head);
+	struct list_node *node2 = NODE_PTR(root, node1->next);
+	struct list_node *node3 = NODE_PTR(root, node2->next);
+
+	if ((root->head == 6 && node1->value != 66) || \
+		(node1->next == 3 && node2->value != 33) || \
+		(node2->next == 5 && node3->value != 55))
+		return 1;
+
+	return 0;
+}
+{% endhighlight %}
+
+That's it. List and consistency checker are ready.
+The next step is to generate a log of memory operations
+made by the application. Such a log we will call store log.
+
+
+### Store log
+
+To get the list of all memory operations we are using pmemcheck logging
+functionality. To turn on logging use `log-stores` parameter:
+
+{% highlight sh %}
+$ valgrind --tool=pmemcheck --log-stores=yes --log-file=store_log.log <your_app> [opt]
+{% endhighlight %}
+
+An output of the above command is a text file with the list of stores, flushes, fences,
+user-defined markers and possibly some other data that can be consumed by pmreorder tool.
+
+Let's look at part of such a store log, and read comments
```

Comment:
Let's look at an excerpt from our example's store log. Make sure to read the
embedded comments.